src/libsyntax/parse/parser/diagnostics.rs

```diff
@@ -916,6 +916,54 @@ impl<'a> Parser<'a> {
         Err(err)
     }
 
+    pub(super) fn expect_semi(&mut self) -> PResult<'a, ()> {
+        if self.eat(&token::Semi) {
+            return Ok(());
+        }
+        let sm = self.sess.source_map();
+        if self.look_ahead(1, |t| {
+            (t == &token::CloseDelim(token::Brace) || token_can_begin_expr(t)) && t.kind != token::Colon
+        }) && [token::Comma, token::Colon].contains(&self.token.kind) {
+            // Likely typo: `,` → `;`, `:` → `;`.
+            match (sm.lookup_line(self.prev_span.lo()), sm.lookup_line(self.token.span.lo())) {
+                (Ok(ref a), Ok(ref b)) if a.line != b.line => {
+                    self.bump();
+                    let sp = self.prev_span;
+                    self.struct_span_err(sp, "expected `;`")
+                        .span_suggestion(
+                            sp,
+                            "change this to `;`",
+                            ";".to_string(),
+                            Applicability::MaybeIncorrect,
+                        ).emit();
+                    return Ok(())
+                }
+                _ => {}
+            }
+        } else if self.look_ahead(0, |t| {
+            (t == &token::CloseDelim(token::Brace) || token_can_begin_expr(t))
+                && t != &token::Semi
+                && t != &token::Pound // Avoid triggering with too many trailing `#` in raw string.
+        }) {
+            // Missing semicolon typo.
+            match (sm.lookup_line(self.prev_span.lo()), sm.lookup_line(self.token.span.lo())) {
+                (Ok(ref a), Ok(ref b)) if a.line != b.line => {
```

Comment:
This could be turned into a method (`fn spans_on_same_line(self: &SourceMap`) that returns `true` if this condition holds. This logic is used twice in this method and `fn unexpected_try_recover` as well (worth checking if other places in this file has it too).

This could probably(?) also be rephrased as:
```rust
self.is_multiline(lo.to(hi))
```